aesd-char-driver/main.c

```diff
@@ -59,6 +60,143 @@ int aesd_release(struct inode *inode, struct file *filp)
     return 0;
 }
 
+
+// Custom llseek function to modify the file pointer position
+// It uses the fixed_size_llseek() function to make the change in the kernel
+static long aesd_adjust_file_offset(struct file *filp , unsigned int write_cmd , unsigned int write_cmd_offset)
+{
+    
+    long ret_val = 0;
+
+    // Iterator
+    int i = 0;
+
+    int ret_status = 0;
+
+    // Store the temporary value of the f_pos before writing to the filp member
+    uint32_t temp_fpos = 0;
+    
+    // Structure to reference the circular buffer
+    struct aesd_dev *dev_offset = filp->private_data;
+
+    PDEBUG("adjust 1\n");
+
+    // Error check 1: If the write command is greater than the allowed number of entries aka 9
+    if (write_cmd > (AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED - 1)){
+        ret_val = -EINVAL;
+        return ret_val;
+    }
+
+    // Error check 2: If the write command offset is greater than allowed offset
+    if (write_cmd_offset >= dev_offset->rw_circular_buffer.entry[write_cmd].size){
+        ret_val = -EINVAL;
+        return ret_val;
+    }
+
+    PDEBUG("adjust 2\n");
+
+    // Locking is necessary since you don't want to operate on a stale copy of the circular buffer
+    ret_status = mutex_lock_interruptible(&dev_offset->rw_mutex_lock);
+
+    // Error check
+    if (ret_status != 0){
+        ret_val = -ERESTARTSYS;
+        return ret_val;
+    }
+
+    PDEBUG("adjust 3\n");
+
+    for (i = 0; i < write_cmd; i++){
+        // Increment the temp file position 
+        temp_fpos += dev_offset->rw_circular_buffer.entry[i].size;
+    }
+
+    PDEBUG("adjust 4\n");
+
+    // Finally, increment it by the required offset in the given command
+    temp_fpos += write_cmd_offset;
+
+    filp->f_pos = temp_fpos;
+
+    PDEBUG("adjust 5\n");
+
+    mutex_unlock(&dev_offset->rw_mutex_lock);
+
+    PDEBUG("adjust 6\n");
+
+    return ret_val;
+}
+
+loff_t aesd_llseek(struct file *filp, loff_t off, int whence)
+{
+    loff_t retval = 0;
+
+    // Mutex return values
+    int ret_status = 0;
+
+    // Store filp structure private data in a local data variable
+    struct aesd_dev *llseek_dev =  filp->private_data;
+
+    ret_status = mutex_lock_interruptible(&aesd_device.rw_mutex_lock);
+
+    // Error check
+    if (ret_status != 0){
+        retval = -ERESTARTSYS;
+        return retval;
+    }
+
+    // Reference: Lecture video
+    // Wrapper and supporting function to implement own llseek function
+    retval = fixed_size_llseek(filp , off , whence , llseek_dev->rw_circular_buffer.total_buff_size);
```

Comment:
@KhyatiSatta how could you remove the need for `fixed_size_llseek` to be in your locking interval?